### PR TITLE
Open hdf files in read mode if read-only is set

### LINKED ISSUE #180: Race condition when reading speclib from two instances
**Describe the bug**
When two alphaDIA instance access the same speclib file (for reading) in the same instance of time, alphabase throws an error (see below).

**Expected behavior**
No error, as the file is just opened for reading in this cases
I think the problem is that files are opened for 'a' here (`hdf.py`)
```
class HDF_File(HDF_Group):
    def __init__():
...
        mode = "w" if delete_existing else "a"
        with h5py.File(file_name, mode):  # , swmr=True):
            pass
```

**Logs**
```
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO - 0:00:00.022445 INFO: Running DynamicLoader
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO - 0:00:00.027390 INFO: Loading .hdf library from /fs/hela_hybrid.small.hdf
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO - 0:00:00.031234 INFO: Traceback (most recent call last):
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/xx/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/alphadia/cli.py", line 333, in run
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     plan = Plan(
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -            ^^^^^
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/alphadia/planning.py", line 126, in __init__
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     self.load_library()
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/alphadia/planning.py", line 205, in load_library
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     spectral_library = dynamic_loader(self.library_path)
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/alphadia/libtransform.py", line 40, in __call__
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     return self.forward(*args)
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -            ^^^^^^^^^^^^^^^^^^^
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/alphadia/libtransform.py", line 121, in forward
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     library.load_hdf(input_path, load_mod_seq=True)
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/alphabase/spectral_library/base.py", line 681, in load_hdf
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     _hdf = HDF_File(
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -            ^^^^^^^^^
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/alphabase/io/hdf.py", line 533, in __init__
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     with h5py.File(file_name, mode):#, swmr=True):
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -          ^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/h5py/_hl/files.py", line 562, in __init__
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     fid = make_fid(name, mode, userblock_size, fapl, fcpl, swmr=swmr)
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "/fs/home/kraken/conda-envs/alphadia-1.6.2/lib/python3.11/site-packages/h5py/_hl/files.py", line 247, in make_fid
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -     fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO -   File "h5py/h5f.pyx", line 102, in h5py.h5f.open
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO - BlockingIOError: [Errno 11] Unable to synchronously open file (unable to lock file, errno = 11, error message = 'Resource temporarily unavailable')
[2024-06-12, 15:19:11 UTC] {ssh.py:526} INFO - 
```

### PR DESCRIPTION
**Pull request by mschwoer**
fixes #180

### CODE REVIEW

**Comment by jalew188** [comment]
Let's check if this is a breaking change for hdf-based mmap @sophiamaedler

File: alphabase/io/hdf.py [hunk 1]
```diff
@@ -498,9 +498,16 @@ def __init__(
...
             "colleagues"
         """
-        mode = "w" if delete_existing else "a"
-        with h5py.File(file_name, mode):  # , swmr=True):
+        if delete_existing:
+            mode = "w"
+        elif read_only:
+            mode = "r"
+        else:
+            mode = "a"
```
Comment (jalew188) [inline alphabase/io/hdf.py:506]: The `truncate` param in the function may become ambiguis

Comment (GeorgWa) [reply alphabase/io/hdf.py:506]: I have the same feeling.
I think the naming is also very ambigous.

**Review by GeorgWa — APPROVED** [review]
LGTM, but more work to do in this module.

**Comment by jalew188** [comment]
MB it is better to re-design this module in a more readable way. We can still use setattr, but it must be implemented by a well named method.

File: alphabase/io/hdf.py (hunk 1 above)
Comment (mschwoer) [reply alphabase/io/hdf.py:511]: not really sure what you mean? is there action required?

Comment (jalew188) [reply alphabase/io/hdf.py:511]: For this function, (how to ref a code block??)
        Args:
            file_name (str): file path.
            read_only (bool, optional): If hdf is read-only. Defaults to True.
            truncate (bool, optional): If existing groups and datasets can be
                truncated (i.e. are overwitten). Defaults to False.
            delete_existing (bool, optional): If the file already exists,
                delete it completely and create a new one. Defaults to False.

**Comment by mschwoer** [comment]
> MB it is better to re-design this module in a more readable way. We can still use setattr, but it must be implemented by a well named method.

definitely this module should be refactored, but this is beyond the scope of this change

**Comment by jalew188** [comment]
> > MB it is better to re-design this module in a more readable way. We can still use setattr, but it must be implemented by a well named method.
> 
> definitely this module should be refactored, but this is beyond the scope of this change

Lets keep the cur version like this then.

**Review by jalew188 — APPROVED** [review]

**Pull request merged**